crates/libcontainer/src/container/builder_impl.rs

```diff
@@ -231,3 +253,89 @@ impl ContainerBuilderImpl {
         Ok(())
     }
 }
+
+struct StdioDescriptors {
+    inner: HashMap<RawFd, RawFd>,
+    outer: HashMap<RawFd, PipeHolder>,
+    _guards: Vec<Closing>,
+}
+
+impl From<&mut StdioDescriptors> for StdioFds {
+    fn from(value: &mut StdioDescriptors) -> Self {
+        StdioFds {
+            stdin: value.outer.remove(&0).and_then(|x| match x {
+                PipeHolder::Writer(x) => Some(x),
+                _ => None,
+            }),
+            stdout: value.outer.remove(&1).and_then(|x| match x {
+                PipeHolder::Reader(x) => Some(x),
+                _ => None,
+            }),
+            stderr: value.outer.remove(&2).and_then(|x| match x {
+                PipeHolder::Reader(x) => Some(x),
+                _ => None,
+            }),
+        }
+    }
+}
+
+fn prepare_stdio_descriptors(fds: &[Fd; 3]) -> Result<StdioDescriptors, LibcontainerError> {
```

Comment:
Can we add comments explaining what this and the `StdioDescriptors` above are used for and how they are used?